### ISSUE TITLE
How to create redux action like models

### ISSUE THREAD

**Issue by rmcarias** [avkonst/hookstate#173]
I have been reading the docs on this library and am loving what I am reading. Basically I need something to handle numerous amounts of form fields; which formik seems to die on.

However, we already have some "redux" like stores and actions. So questions:
1. How do I share the state object without passing it down so many layers or props (I see most of your examples, especially for scoped, it's all about passing down the prop. But that can get into real bad prop drilling)?
2. If need to create actions with side-effects, is there a concept of dispatch or is it just plain methods I can add to the model when I create the global store?

**Comment by rmcarias** [comment]
Thanks for the quick response @avkonst . I think you covered it in terms of having custom accessible methods (hookstateplugin). Can I have a method in there that is async (alongside the onSet, etc..)?

As for sharing via global state, is still performant if it's not used as a scoped field (In your docs it mentions that for a large number of fields to use scoped state; but not sure if I can do that with a global share)?

**Comment by avkonst** [comment]
"Can I have a method in there that is async (alongside the onSet, etc..)?"
No, you can not as it is synchronous as the underlaying `setState` method from `React.useState`. However, nothing stops you to spawn async calls there, which return Promises and complete eventually.

"As for sharing via global state, is still performant if it's not used as a scoped field"

If you use pure global state, it will rerender only affected components, however the algorithm to find affected components would be `O(n)` scalability instead of nearly `O(1)` offered by the scoped state. If you do not have more than few hundreds of subscribed components and changes are not frequent, the difference would be minimal.

In general global state is used for things like `user state`, which is accessed by many components too far from each other in the tree, eg. navbar user icon and a deeply nested component in the main section.

When you are talking about a component holding a transient / temporary form state, which has all of the nested components which use the state relatively close to each other, you are better to use props instead of a global variable. It is cleaner and it is better performance.

And you can combine the global state with scoped state too. Your form state can be created and stored in a global variable. Your form component can use it as a global state and pass nested states as props to children components, which would use props as the scoped state. In this case you will get all of the benefits from both worlds, but props drilling would be required at least for the children component of the form state parent component.

**Comment by avkonst** [comment]
I can not do await. Just return a promise, but it also means you can have multiple incomplete spawned properties running in parallel, if your set action is called multiple times without awaiting the previous Promises.

There is also a design technique which allows to chain spawned and not awaited Promises in a single stack, which would drive execution of Promises one by one, awaiting one by one, but it is out of scope of this package.
